flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java

```diff
@@ -0,0 +1,319 @@
+			ObjectPath newPath = new ObjectPath(tablePath.getDatabaseName(), newTableName);
+
+			if (tableExists(newPath)) {
+				throw new TableAlreadyExistException(catalogName, newPath);
+			} else {
+				tables.put(newPath, tables.remove(tablePath));
+			}
+		} else if (!ignoreIfNotExists) {
+			throw new TableNotExistException(catalogName, tablePath);
+		}
+	}
+
+	@Override
+	public List<String> listAllTables(String databaseName) throws DatabaseNotExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName), "dbName cannot be null or empty");
+
+		if (!databaseExists(databaseName)) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+
+		return tables.keySet().stream()
+			.filter(k -> k.getDatabaseName().equals(databaseName)).map(k -> k.getObjectName())
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public List<String> listTables(String databaseName) throws DatabaseNotExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName), "dbName cannot be null or empty");
+
+		if (!databaseExists(databaseName)) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+
+		return tables.keySet().stream()
+			.filter(k -> k.getDatabaseName().equals(databaseName))
+			.filter(k -> (tables.get(k) instanceof CatalogTable)).map(k -> k.getObjectName())
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public CommonTable getTable(ObjectPath tablePath) throws TableNotExistException {
+		checkArgument(tablePath != null);
+
+		if (!tableExists(tablePath)) {
+			throw new TableNotExistException(catalogName, tablePath);
+		} else {
+			return tables.get(tablePath).copy();
+		}
+	}
+
+	@Override
+	public boolean tableExists(ObjectPath tablePath) {
+		return tablePath != null && databaseExists(tablePath.getDatabaseName()) && tables.containsKey(tablePath);
+	}
+
+	// ------ views ------
+
+	@Override
+	public List<String> listViews(String databaseName) throws DatabaseNotExistException {
+		checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName), "dbName cannot be null or empty");
+
+		if (!databaseExists(databaseName)) {
+			throw new DatabaseNotExistException(catalogName, databaseName);
+		}
+
+		return tables.keySet().stream()
+			.filter(k -> k.getDatabaseName().equals(databaseName))
+			.filter(k -> (tables.get(k) instanceof CatalogView)).map(k -> k.getObjectName())
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public void createView(ObjectPath viewPath, CatalogView view, boolean ignoreIfExists)
+		throws TableAlreadyExistException, DatabaseNotExistException {
+		checkArgument(viewPath != null);
+		checkArgument(view != null);
+
+		if (!databaseExists(viewPath.getDatabaseName())) {
+			throw new DatabaseNotExistException(catalogName, viewPath.getDatabaseName());
+		}
+
+		if (tableExists(viewPath)) {
+			if (!ignoreIfExists) {
+				throw new TableAlreadyExistException(catalogName, viewPath);
+			}
+		} else {
+			tables.put(viewPath, view.copy());
+		}
```

Comment:
The implementations of createView() and createTable() are same. It's better to deduplicate the code. Same for alterView() and alterTable(). Again, I think maybe we can provide an api for both table and view, i.e., createTable(ObjectPath tablePath, CommonTable table, boolean ignoreIfExists) and alterTable(ObjectPath tablePath, CommonTable newTable, boolean ignoreIfNotExists). Table here means table and view.

Reply:
I can provide a private helper method that does the logic and both createView() and createTable() can call the helper method. Makes sense? The implementation here are the same because of inmemory nature. In general, they would be different, and CatalogTable and CatalogView are two different things. Thoughts?